File: dp/longest_repeating_subsequence.py
# User function Template for python3


class Solution:
    def LongestRepeatingSubsequence(self, str):
        # Code here
        n = len(str)
        dp = [[0 for _ in range(n + 1)] for _ in range(n + 1)]
        for i in range(n - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                if i != j and str[i] == str[j]:
                    dp[i][j] = dp[i + 1][j + 1] + 1
                else:
                    dp[i][j] = max(dp[i][j + 1], dp[i + 1][j])

        return dp[0][0]


# {
# Driver Code Starts
# Initial Template for Python 3

if __name__ == "__main__":
    T = int(input())
    for i in range(T):
        str = input()
        ob = Solution()
        ans = ob.LongestRepeatingSubsequence(str)
        print(ans)

# } Driver Code Ends

File: longest_repeating_subsequence.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
    int dp[1001][1001];
    int solve(const string&s, int i, int j) {
        if (j == s.size()) return 0;
        if (dp[i][j] != -1) return dp[i][j];
        int& ans = dp[i][j];
        ans = 0;
        ans = solve(s, i, j+1);
        for(int k = i; k < j; ++k) {
            if (s[k] == s[j]) {
                return ans = max(ans, solve(s, k+1, j+1) + 1);
            } else {
                ans = max(ans, solve(s, k+1, j));
            }
        }
        return ans;
    }
public:
	int LongestRepeatingSubsequence(string str){
		   // Code here
	    memset(dp, -1, sizeof(dp));
	    return solve(str, 0, 1);
	}
};

//{ Driver Code Starts.
int main(){
	int tc;
	cin >> tc;
	while(tc--){
		string str;
		cin >> str;
		Solution obj;
		int ans = obj.LongestRepeatingSubsequence(str);
		cout << ans << "\n";
	}
	return 0;
}
// } Driver Code Ends
